oeps/oep-0033-bp-code-autoformatting.rst

```diff
@@ -0,0 +1,127 @@
+make for more productive code reviews and waste less time on inconsequential
+details.
+
+Decision
+--------
+
+We will select automatic code formatters for all of our major languages. These
+formatters will have both a lint-mode (that will be run in continuous
+integration to ensure consistent formatting) and an autoformat mode (which
+can be integrated with editors to free developers from having to think about
+code formatting). These formatting standards will be applied across all
+Open edX-authored repositories.
+
+Selected tools should, ideally, require minimal configuration. Any required
+configuration should be added to `edx/edx-lint`_ in order to allow it to
+be centralized and standardized across repositories in the Open edX ecosystem.
+
+The current recommended formatters are:
+
+    - Python: `black`_
+    - Javascript: `prettier`_
+    - CSS/SASS: `prettier`_
+    - HTML: `prettier`_
+    - JSON: `prettier`_
+    - PHP: `prettier-php`_
+
+.. _black: https://github.com/ambv/black
+.. _prettier: https://prettier.io/
+.. _prettier-php: https://github.com/prettier/plugin-php
+.. _`edx/edx-lint`: https://github.com/edx/edx-lint
+
+Formatting should exposed via the standard ``Makefile`` commands
+``format`` (which should autoformat all code in the repository)
+and ``check-format`` (which should validate that all code in the
+repository is formatted). ``check-format`` should also be called
+as part of ``make quality``.
+
+When bulk-formatting the entire repo for the first time, all changes should
+be done in as few commits as possible (ideally, only one). Then, the
+commit hashes should be added to a ``.git-blame-ignore-revs`` file.
+This will allow the `git hyper-blame`_ tool to ignore them by default,
+to preserve the easy ability to see who changed particular file lines.
+
+.. _git hyper-blame: https://commondatastorage.googleapis.com/chrome-infra-docs/flat/depot_tools/docs/html/git-hyper-blame.html
+
+Consequences
+------------
+
+There will be commits to all Open edX repositories that cause large-scale
+reformatting without functional changes. This will muddy the blame-file
+history. Users with auto-formatters configured in their editors working
+in a repo that is not yet OEP-33 compliant will need to disable their
+autoformatter, or risk large-scale changes that are unrelated to their
+current tasks. All Open edX code will be formatted consistently and
+unambiguously. Developers will not have arguments about
+code layout.
+
+References
+----------
+
+`gofmt`_ was the first code-autoformatter to gain wide-scale acceptance.
+
+.. _gofmt: https://blog.golang.org/go-fmt-your-code
+
+`Black editor integration`_
+`Prettier editor integration`_
+
+.. _Black editor integration: https://github.com/ambv/black#editor-integration
+.. _Prettier editor integration: https://prettier.io/docs/en/editors.html
+
+An example ``tox.ini``:
+
+.. code-block::
+
+    [testenv:format]
+    basepython =
+        python3.6
+    deps=black
```

Comment:
From an OEP-18 perspective, this should use a requirements file generated by `pip-compile` instead (so we get version consistency and regular updates).  I realize this is just an example, but I don't want to encourage people to copy this pattern verbatim.